src/core/ext/filters/client_channel/lb_policy/weighted_round_robin/weighted_round_robin.cc

```diff
@@ -0,0 +1,955 @@
+    OnBackendMetricReport(const BackendMetricData& backend_metric_data) {
+  weight_->MaybeUpdateWeight(backend_metric_data.qps,
+                             backend_metric_data.cpu_utilization);
+}
+
+//
+// WeightedRoundRobin::WeightedRoundRobinSubchannelData
+//
+
+WeightedRoundRobin::WeightedRoundRobinSubchannelData::
+    WeightedRoundRobinSubchannelData(
+        SubchannelList<WeightedRoundRobinSubchannelList,
+                       WeightedRoundRobinSubchannelData>* subchannel_list,
+        const ServerAddress& address, RefCountedPtr<SubchannelInterface> sc)
+    : SubchannelData(subchannel_list, address, std::move(sc)),
+      weight_(static_cast<WeightedRoundRobin*>(subchannel_list->policy())
+                  ->GetOrCreateWeight(address.address())) {
+  // Start OOB watch if configured.
+  WeightedRoundRobin* p =
+      static_cast<WeightedRoundRobin*>(subchannel_list->policy());
+  if (p->config_->enable_oob_load_report()) {
+    subchannel()->AddDataWatcher(
+        MakeOobBackendMetricWatcher(p->config_->oob_reporting_period(),
+                                    std::make_unique<OobWatcher>(weight_)));
+  }
+}
+
+void WeightedRoundRobin::WeightedRoundRobinSubchannelData::
+    ProcessConnectivityChangeLocked(
+        absl::optional<grpc_connectivity_state> old_state,
+        grpc_connectivity_state new_state) {
+  WeightedRoundRobin* p =
+      static_cast<WeightedRoundRobin*>(subchannel_list()->policy());
+  GPR_ASSERT(subchannel() != nullptr);
+  // If this is not the initial state notification and the new state is
+  // TRANSIENT_FAILURE or IDLE, re-resolve.
+  // Note that we don't want to do this on the initial state notification,
+  // because that would result in an endless loop of re-resolution.
+  if (old_state.has_value() && (new_state == GRPC_CHANNEL_TRANSIENT_FAILURE ||
+                                new_state == GRPC_CHANNEL_IDLE)) {
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_lb_wrr_trace)) {
+      gpr_log(GPR_INFO,
+              "[WRR %p] Subchannel %p reported %s; requesting re-resolution", p,
+              subchannel(), ConnectivityStateName(new_state));
+    }
+    p->channel_control_helper()->RequestReresolution();
+  }
+  if (new_state == GRPC_CHANNEL_IDLE) {
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_lb_wrr_trace)) {
+      gpr_log(GPR_INFO,
+              "[WRR %p] Subchannel %p reported IDLE; requesting connection", p,
+              subchannel());
+    }
+    subchannel()->RequestConnection();
+  } else if (new_state == GRPC_CHANNEL_READY) {
+    // If we transition back to READY state, restart the blackout period.
+    // Note that we cannot guarantee that we will never receive
+    // lingering callbacks for backend metric reports from the previous
+    // connection after the new connection has been established, but they
+    // should be masked by new backend metric reports from the new
+    // connection by the time the blackout period ends.
+    weight_->ResetNonEmptySince();
```

Comment:
Why don't we do this when the channel moves out of READY and also simply invalidate all weight information at all? As you mentioned since there can always be an extreme timing issue w.r.t. synchronization, it looks to me that resetting values at the earliest is safer and easier to maintain. Also all weight states (weight and two timestamps) become invalid will never be used as soon as a subchannel becomes non-READY so we don't need to keep them, correct?